### ISSUE TITLE
validate() heuristic for detecting path vs graph is buggy

### ISSUE THREAD

**Issue by Zezombye** [RDFLib/pySHACL#132]
In load.py, load_from_source():

```py
    elif isinstance(source, str):
        if is_windows and source.startswith('file:///'):
            public_id = source
            source_is_file = True
            filename = source[8:]
        elif not is_windows and source.startswith('file://'):
            public_id = source
            source_is_file = True
            filename = source[7:]
        elif source.startswith('http:') or source.startswith('https:'):
            public_id = source
            try:
                resp, rdf_format = get_rdf_from_web(source)
            except HTTPError:
                if is_imported_graph:
                    return g
                else:
                    raise
            if rdf_format == 'graph':
                source = resp
                source_is_graph = True
            else:
                filename = resp.geturl()
                source = resp.fp
                source_was_open = False
                source_is_open = True
        else:
            first_char = source[0]
            if is_windows and (first_char == '\\' or (len(source) > 3 and source[1:3] == ":\\")):
                source_is_file = True
                filename = source
            elif first_char == '/' or source[0:3] == "./":
                source_is_file = True
                filename = source
            elif (
                first_char == '#'
                or first_char == '@'
                or first_char == '<'
                or first_char == '\n'
                or first_char == '{'
                or first_char == '['
            ):
                # Contains some JSON or XML or Turtle stuff
                source_is_file = False
            elif len(source) < 140:
                source_is_file = True
                filename = source
        # TODO: Do we still need this? Not sure why this was added, but works better without it
        #  if public_id and not public_id.endswith('#'):
        #     public_id = "{}#".format(public_id)
        if not source_is_file and not source_is_open and isinstance(source, str):
            # source is raw RDF data.
            source = source.encode('utf-8')
            source_is_bytes = True
```

There is a bug in `or source[0:3] == "./"` which never evaluates to true, as such the heuristic ends up checking if the path is less than 140 chars... I lost quite a bit of time debugging when I switched my rules to a longer path.

The whole idea of heuristically checking if a string is a file or a graph is flawed however. Here it could also bug out on graphs less than 140 characters that start with \r\n (on windows). Files can also begin with a "[". Etc.

The solution would be either:

- Test if the string is a path via os.path.exists(), however if the file doesn't exist then it will also raise a confusing "syntax error" instead of "file not found";
- Have separate arguments/functions for graphs and files;
- Only take a graph and have the user read the file by themselves (which is easy with python).

**Comment by ashleysommer** [comment]
Hi @Zezombye 
> The whole idea of heuristically checking if a string is a file or a graph is flawed however

I agree. Its not a good way of doing it, but one of the original design goals of PySHACL was to make it the graph take in either a filename, web address (http),or raw data. That may not have been a good idea at the time, but it is feature of the tool now, regardless.

You've deduced correctly that the system attempts to apply some crude heuristics to detect when the string is a file path, and they generally work fine for the majority of users. However, as you pointed out, one of the more important ones is bugged.

We're putting out a new version of PySHACL very soon (this weekend), so for a quick fix, I'll correct that heuristic to read `or source[0:2] == "./"` as it should.

I will put some thought into this, and try to come up with a better way of doing it (eg, file.exists() detection as you mentioned).

**Comment by Zezombye** [comment]
Thanks! A better heuristic would also be to check for \n in the string (if yes, then it is a file), and doing strip() before getting the first char.

**Comment by ashleysommer** [comment]
@Zezombye  this small fix was released in v0.19.0.
The larger fix for this will be released in v1.0, which will be the next big release and will allow breaking changes (eg, we could disallow raw source text to be passed in).
